### PR TITLE
Automated test to verify AMD GPU's can be used in Data Science Pipelines

### PR DESCRIPTION
**Pull request by diegolovison**
Tested on Azure with mi300x

```
sh run_robot_test.sh --set-urls-variables true --extra-robot-args '-L DEBUG -i AMD-GPUs' --skip-oclogin
```

```
Verify Pipeline Tasks Run On AMD GPU Nodes Only When Tolerations A... | PASS |
------------------------------------------------------------------------------
Tests.Data Science Pipelines.Data-Science-Pipelines-Gpu :: Basic a... | PASS |
1 test, 1 passed, 0 failed
==============================================================================
Tests.Data Science Pipelines                                          | PASS |
1 test, 1 passed, 0 failed
==============================================================================
Tests                                                                 | PASS |
1 test, 1 passed, 0 failed
==============================================================================
Output:  /home/dlovison/github/rhods/ods-ci/ods_ci/test-output/ods-ci-2024-10-15-16-01-HEMIY4BvUv/output.xml
XUnit:   /home/dlovison/github/rhods/ods-ci/ods_ci/test-output/ods-ci-2024-10-15-16-01-HEMIY4BvUv/xunit_test_result.xml
Log:     /home/dlovison/github/rhods/ods-ci/ods_ci/test-output/ods-ci-2024-10-15-16-01-HEMIY4BvUv/log.html
Report:  /home/dlovison/github/rhods/ods-ci/ods_ci/test-output/ods-ci-2024-10-15-16-01-HEMIY4BvUv/test_report.html

```

### CODE REVIEW

File: ods_ci/tests/Tests/1100__data_science_pipelines/1105__data-science-pipelines-gpu.robot [hunk 1]
```diff
@@ -4,46 +4,62 @@
...
+Verify Pipeline Tasks Run On Nvidia GPU Nodes Only When Tolerations Are Added
     [Documentation]    Runs a pipeline that tests GPU availability according to GPU tolerations in pipeline tasks:
     ...    - One task should not have GPUs available, as we don't add the GPU tolerations
     ...    - Another task should have GPUs available, as we add the GPU tolerations
     [Tags]    Tier1    Resources-GPU    NVIDIA-GPUs
 
+    Verify Pipeline Tasks Run On GPU Nodes Only When Tolerations Are Added    ${PIPELINE_NVIDIA_GPU}
+
+# robocop: off=too-long-test-case
+Verify Pipeline Tasks Run On AMD GPU Nodes Only When Tolerations Are Added
```
[the remark this answers is not in the document]
Comment (diegolovison) [reply ods_ci/tests/Tests/1100__data_science_pipelines/1105__data-science-pipelines-gpu.robot:30]: If I add one more it complains as well

File: ods_ci/tests/Resources/Files/pipeline-samples/v2/cache-disabled/gpu/pytorch/pytorch_amd_gpu_availability.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
+from kfp import compiler, dsl, kubernetes
+from kfp.dsl import PipelineTask
+
+#  Runtime: Pytorch with ROCm and Python 3.9 (UBI 9)
+common_base_image = (
+    "quay.io/modh/runtime-images@sha256:76d2a5d9ca0bb35c4d9962a7ee6798672afd3a5c7d226d87ef65f952e7cb93cf"
+)
+
+
+def add_gpu_toleration(task: PipelineTask, accelerator_type: str, accelerator_limit: int):
```
[the remark this answers is not in the document]
Comment (jgarciao) [reply ods_ci/tests/Resources/Files/pipeline-samples/v2/cache-disabled/gpu/pytorch/pytorch_amd_gpu_availability.py:10]: The toleration is added to the pipeline task **pod** (it's not a taint in the node). The pod will be deleted in the Suite Teardown when the project is deleted

[the remark this answers is not in the document]
Comment (bdattoma) [reply ods_ci/tests/Resources/Files/pipeline-samples/v2/cache-disabled/gpu/pytorch/pytorch_amd_gpu_availability.py:10]: AH right, sorry my bad

File: ods_ci/tests/Tests/1100__data_science_pipelines/1105__data-science-pipelines-gpu.robot [hunk 3]
```diff
@@ -4,46 +4,65 @@ Resource            ../../Resources/RHOSi.resource
...
     ...    pipeline_name=pytorch-verify-gpu-availability
     ...    pipeline_description=Verifies GPU availability in tasks when using tolerations
-    ...    pipeline_package_path=${PIPELINE_GPU_AVAILABILITY_FILEPATH}
+    ...    pipeline_package_path=${pipeline_package_path}
     ...    pipeline_run_name=pytorch-verify-gpu-availability-run
 
     DataSciencePipelinesBackend.Wait For Run Completion And Verify Status
     ...    namespace=${PROJECT}    username=${TEST_USER.USERNAME}    password=${TEST_USER.PASSWORD}
-    ...    pipeline_run_id=${pipeline_run_id}    pipeline_run_timeout=240
+    ...    pipeline_run_id=${pipeline_run_id}    pipeline_run_timeout=600
```
Comment (jgarciao) [inline ods_ci/tests/Tests/1100__data_science_pipelines/1105__data-science-pipelines-gpu.robot:58]: Is this change needed? The pipeline usually runs very fast, but if there are no GPUs available (e.g. because they are used by another pod) we want it to fail soon

**Review by jgarciao — CHANGES_REQUESTED** [review]

**Review by jgarciao — APPROVED** [review]

File: ods_ci/tests/Resources/Files/pipeline-samples/v2/cache-disabled/gpu/pytorch/pytorch_amd_gpu_availability.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+)
+def verify_gpu_availability(gpu_toleration: bool):
+    import torch
+
+    cuda_available = torch.cuda.is_available()
+    device_count = torch.cuda.device_count()
+    print("------------------------------")
+    print("GPU availability")
+    print("------------------------------")
+    print(f"cuda available: {cuda_available}")
```
Comment (tarukumar) [inline ods_ci/tests/Resources/Files/pipeline-samples/v2/cache-disabled/gpu/pytorch/pytorch_amd_gpu_availability.py:28]: I think it make sense to replace cuda with rocm

Comment (diegolovison) [reply ods_ci/tests/Resources/Files/pipeline-samples/v2/cache-disabled/gpu/pytorch/pytorch_amd_gpu_availability.py:28]: PyTorch does not have a separate device string for ROCm, which is why you use 'cuda' to refer to an AMD GPU under ROCm.

Comment (tarukumar) [reply ods_ci/tests/Resources/Files/pipeline-samples/v2/cache-disabled/gpu/pytorch/pytorch_amd_gpu_availability.py:28]: I know that and that is why i am suggesting to change the print statement with rocm instead of cuda so that people can differentiate easily.

**Review by bdattoma — APPROVED** [review]

File: ods_ci/tests/Resources/Files/pipeline-samples/v2/cache-disabled/gpu/pytorch/pytorch_amd_gpu_availability.py (hunk 4 above)
Comment (diegolovison) [reply ods_ci/tests/Resources/Files/pipeline-samples/v2/cache-disabled/gpu/pytorch/pytorch_amd_gpu_availability.py:28]: imho, if we are getting "cuda" we should print "cuda". in this case, we don't "hide" any detail.
also, having cuda string for rocm is super confusing.
